### PR TITLE
Fix 10532 to change --api-enable-cors to string.

### LINKED ISSUE #10532: hardcoded * as Access-Control-Allow-Origin value
Some implementations of CORS doesn't allow `*` as a value for Access-Control-Allow-Origin header.
Currently this  value is hardcoded [in line 1223 of server.go](https://github.com/docker/docker/blob/master/api/server/server.go#L1223)
Parameter --api-enable-cors  could be a string, instead of bool, allowing to pass an arbitrary value to "Access-Control-Allow-Origin" header falling back to `*` as default.

### PR DESCRIPTION
**Pull request by resouer**
Fix https://github.com/docker/docker/issues/10532

cors heads was hard coded as \* which is 
1. not safe
2. can not modify to specify domain

Signed-off-by: harry zhang resouer@163.com

### CODE REVIEW

**Comment by SvenDowideit** [comment]
I think this needs documentation in the man pages and the cli

**Comment by resouer** [comment]
@SvenDowideit I have updated docs and fix CI. Please review it.

File: docs/sources/reference/commandline/cli.md [hunk 1]
```diff
@@ -69,7 +69,7 @@ expect an integer, and they can only be specified once.
     A self-sufficient runtime for linux containers.
 
     Options:
-      --api-enable-cors=false                    Enable CORS headers in the remote API
+      --api-enable-cors=false                    CORS headers, true or * to allow all
```
Comment (tiborvass) [inline docs/sources/reference/commandline/cli.md]: `="false"`

**Comment by tiborvass** [comment]
@resouer this PR would need to deprecate `--api-enable-cors` in favor of `--api-cors-headers` or something like that.

Ping @jfrazelle @crosbymichael

**Comment by resouer** [comment]
@tiborvass I meant to do like that, but after I googled `api-enable-cors`, I found it has been referenced by a lot of project in dockerd start command, so I kept that to avoid corrupting them. Any idea about that?

**Comment by tiborvass** [comment]
@resouer we should still keep it for backwards compatibility, but if you want your fix, I personally think it makes more sense semantically to use a new flag, since it's not a boolean anymore it's not about "enabling", it's the **actual** CORS header that we're setting.

So here's what I suggest:
* keep `--api-enable-cors` a boolean, as it is today
* add another flag `--api-cors-header` that if non-empty, implies `--api-enable-cors=true`, and sets the CORS header according to the string passed to the flag.
* (optional) add a deprecation warning message for `--api-enable-cors` that incites to use the new `--api-cors-header` flag.

**Comment by resouer** [comment]
@tiborvass can't agree more with you, updating the code.

**Comment by tiborvass** [comment]
@resouer thanks, but keep in mind that this is subject to Design review, and I only voiced my opinion :) Introducing a new flag is not something that's easily done. So I'll need approval from other maintainers before you can be 100% sure :)

Ping @crosbymichael @jfrazelle @LK4D4 @icecrime @vieux

**Comment by resouer** [comment]
IMO, +1 for keep the old flag. Waiting to get others' feedback

**Comment by resouer** [comment]
@tiborvass I have updated the code & docs which left `--api-enable-cors` no changed. Please review.

**Comment by resouer** [comment]
@tiborvass @LK4D4 @jfrazelle  Any agreement on this design?

**Comment by tiborvass** [comment]
@crosbymichael @jfrazelle @LK4D4 @icecrime @vieux 
Here's what I suggested: https://github.com/docker/docker/pull/10586#issuecomment-73347941

+1 the design or give alternatives, thanks!

**Comment by icecrime** [comment]
+1 on @tiborvass suggested design!

**Comment by vieux** [comment]
+1 on the design as well and +1 for the warning.

**Comment by jfrazelle** [comment]
+1

**Comment by shykes** [comment]
design LGTM

**Comment by resouer** [comment]
ping @tiborvass  any step further?

**Comment by jfrazelle** [comment]
this needs a rebase and to be added to the 1.18 v api docs not 1.17

**Comment by resouer** [comment]
@jfrazelle rebase done. Please review it.

File: daemon/config.go [hunk 2]
```diff
@@ -68,7 +69,8 @@ func (config *Config) InstallFlags() {
 	flag.BoolVar(&config.EnableSelinuxSupport, []string{"-selinux-enabled"}, false, "Enable selinux support")
 	flag.IntVar(&config.Mtu, []string{"#mtu", "-mtu"}, 0, "Set the containers network MTU")
 	flag.StringVar(&config.SocketGroup, []string{"G", "-group"}, "docker", "Group for the unix socket")
-	flag.BoolVar(&config.EnableCors, []string{"#api-enable-cors", "-api-enable-cors"}, false, "Enable CORS headers in the remote API")
+	flag.BoolVar(&config.EnableCors, []string{"#api-enable-cors", "#-api-enable-cors"}, false, "Enable CORS headers in the remote API")
```
Comment (tiborvass) [inline daemon/config.go]: @resouer Please mention in the description that it is deprecated by `--api-cors-header`.

**Comment by tiborvass** [comment]
@resouer something's wrong with your branch :( Bad rebase?

**Comment by resouer** [comment]
@tiborvass  Sorry for a bad rebase, it is fixed now

**Comment by resouer** [comment]
@tiborvass Is this shippable now? Just not want to rebase it again ...

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
Ping @jfrazelle @vieux

**Comment by jfrazelle** [comment]
needs a rebase

**Comment by resouer** [comment]
@jfrazelle @tiborvass  Rebase done :heart:  Go go go!

**Comment by jfrazelle** [comment]
LGTM ping @vieux just for final approval

**Comment by vieux** [comment]
LGTM

**Comment by alfonsodev** [comment]
Do you think it would be possible to expose --api-cors-header parameter in Docker for Mac and Docker for Windows `Preferences -> Advanced` menu  ?
screen shot 2016-06-20 at 19 39 50  
Thanks!

**Comment by thaJeztah** [comment]
@alfonsodev I think there's plans to expose more settings through the Docker for Mac UI, however, the Docker for Mac UI is managed by a different team; perhaps you can start a topic on the Docker for Mac forums? https://forums.docker.com/c/docker-for-mac

**Pull request merged**